File: JavaScript/Easy/MajorityElement.js
/**
 * @param {number[]} nums
 * @return {number}
 */
 var majorityElement = function(nums) {
    
  let map = {};
  
  for (let i = 0; i < nums.length; i++) {
      map[nums[i]] ? map[nums[i]]++ : map[nums[i]] = 1;
      if (map[nums[i]] > nums.length/2) return nums[i];
  }
  
};

/*
  
  Create a map that tracks the total number of occurences of each number
  
  Time Complexity: O(n)
  Space Complexity: O(n)
  
  Where n is the length of the nums array

*/

File: C#/Easy/MajorityElement.cs
public class Solution
{
    public int MajorityElement(int[] nums)
    {
        int n = nums.Length;
        int result = 0;
        for (int i = 0; i < 32; i++)
        {
            int bit = 1 << i;

            int bitCount = 0;
            foreach (int num in nums)
            {
                if ((num & bit) == bit) bitCount++;
            }

            if (bitCount > n / 2) result |= bit;
        }
        
        return result;
    }
}

/*

    Bit manipulation

    Time: O(nlogm)
    Space: O(1)

    Where n is the length of nums and m is the size of the largest possible value in nums.

    ------------------------------------------------------------------------
    
    For each number in nums, count the occurrence of each bit, if it's greater than n/2, then it should be set in the result.

*/

